### PR TITLE
Fixes #36342 - Move JS scripts to the layout's Body

### PR DESCRIPTION
**Pull request by Ron-Lavi**
for better dependencies loading and performance.

- It fixes the console error of loading the noVNC library in the head.

in addition, tested the following flow with this commit vs develop branch:
- Start in dashboard
- go to /hosts
- click on hostname which will take you to the new host page (load a full-React page)
- go to /foreman_tasks/tasks (testing a plugin)
- go to Audits (core page - fully in React) 

The results from this commit:
Screenshot from 2023-05-01 15-42-22

The results from develop branch:
Screenshot from 2023-05-01 15-41-21

The main difference in scripts loading was almost twice faster!!

### CODE REVIEW

**Comment by Ron-Lavi** [comment]
ok to test

**Comment by Ron-Lavi** [comment]
[test katello]

**Comment by MariaAga** [comment]
It looks like its safe and does break anything from what I see, @ekohl @evgeni any one against this?

File: app/views/layouts/base.html.erb [hunk 1]
```diff
@@ -32,27 +32,24 @@
...
     <%= yield(:head) %>
   </head>
 
   <body class="pf-m-redhat-font">
-    <!--[if lte IE 9]>
-      <%= _('You are using an unsupported browser.') %>
-    <![endif]-->
+   
     <% if User.current %>
       <% cache(TopbarSweeper.fragment_name, :skip_digest => true) do %>
```
Comment (ekohl) [inline app/views/layouts/base.html.erb:57]: Does this cache influence things? I've never looked at the caching structure.

Comment (Ron-Lavi) [reply app/views/layouts/base.html.erb:57]: When adding it to the cache it seemed like the layout was rendered faster and there was a smoother transition between pages and components loading, but I think it's not required and scripts can also live in the body without the cache.

I was putting it here since the `<%= mount_react_app %>` was there

if you want, I could compare the performance with/out the cache

Comment (ekohl) [reply app/views/layouts/base.html.erb:57]: I was more thinking about cache invalidation etc. It probably doesn't matter here, but if you call the same fragment from multiple pages it may end up with invalid caches.

Comment (ekohl) [reply app/views/layouts/base.html.erb:57]: I do suspect it can cause problems because the `fragment_name` is defined as:
https://github.com/theforeman/foreman/blob/2dc52832034fb8dc289819ebc04f26af13902d33/app/services/topbar_sweeper.rb#L12-L14

So this is cached for each user. Then `mount_react_app` is defined as:

https://github.com/theforeman/foreman/blob/2dc52832034fb8dc289819ebc04f26af13902d33/app/helpers/layout_helper.rb#L2-L8

All of that data is not controller dependent, so it's fine.

However, I am worried that the other content could be problematic. In particular `javascripts` and `inline_javascripts` which are yielded. If we include controller specific JS, such as `app/assets/javascripts/hosts.js` on the hosts page then I'd think it can get cached across multiple pages thus making it can become unreliable. The order in which you navigate pages will determine if it works or not.

Another thing is cache efficiency. None of the things you include here are not user dependent. So including it in a per-user cache will mean you transfer more data in & out of the cache. You take up more cache space, meaning that under pressure the cache is more likely to evict data.

I don't know about the cost of calculating those JS tags, but I'm willing to bet that it's very cheap so caching it doesn't bring any performance benefit and perhaps even bring a net-negative.

If we find that it is expensive to calculate, then it deserves its own cache entry. One which is not bound to a specific user.

In conclusion: please move it out of the cached part.

Comment (Ron-Lavi) [reply app/views/layouts/base.html.erb:57]: Thanks for the investigation, moved them out of the cache block,
and the pages seem to load faster than before.

File: app/views/layouts/base.html.erb [hunk 2]
```diff
@@ -32,26 +32,26 @@
...
-
+    
     <%= yield(:head) %>
   </head>
 
   <body class="pf-m-redhat-font">
-    <!--[if lte IE 9]>
+    <!--[if IE]>
       <%= _('You are using an unsupported browser.') %>
+      <%= _('See list of supported browsers: %s') % 'https://theforeman.org/manuals/nightly/index.html#3.1.4BrowserCompatibility' %>
```
Comment (ekohl) [inline app/views/layouts/base.html.erb]: We have a helper to link to documentation, which takes the version and branding into account.
```suggestion
      <%= _('See list of supported browsers: %s') % documentation_url('3.1.4BrowserCompatibility') %>
```

File: app/views/layouts/base.html.erb [hunk 3]
```diff
@@ -32,26 +32,26 @@
...
-      <%= yield(:inline_javascripts) %>
-    </script>
-
+    
     <%= yield(:head) %>
   </head>
 
   <body class="pf-m-redhat-font">
-    <!--[if lte IE 9]>
+    <!--[if IE]>
```
Comment (ekohl) [inline app/views/layouts/base.html.erb]: I wanted to look into this. https://learn.microsoft.com/en-us/previous-versions/windows/internet-explorer/ie-developer/compatibility/hh801214(v=vs.85) states that IE 10 ignores this so these statements are equal. Given they are equal, I'd personally not make the change but I'll leave it up to you.

File: app/views/layouts/base.html.erb [hunk 4]
```diff
@@ -32,26 +32,26 @@
...
+      <%= javascript_include_tag *webpack_asset_paths('foreman-vendor', :extension => 'js') %>
+      <%= javascript_include_tag *webpack_asset_paths('vendor', :extension => 'js') %>
+      <%= javascript_include_tag *webpack_asset_paths('bundle', :extension => 'js') %>
+      <%= javascript_include_tag 'application' %>
+      <%= webpacked_plugins_with_global_js %>
+      <%= webpack_dev_server %>
+      <%= yield(:javascripts) %>
+      <script type="text/javascript">
+        <%= yield(:inline_javascripts) %>
+      </script>
```
Comment (ekohl) [inline app/views/layouts/base.html.erb]: Are these pages not needed for anonymous pages (such as the login page)? I'd think that you want to include them unconditionally and thus move it above `<% if User.current %>`.

**Review by ekohl — CHANGES_REQUESTED** [review]

File: app/views/layouts/base.html.erb (hunk 4 above)
Comment (Ron-Lavi) [reply app/views/layouts/base.html.erb]: I don't think they are needed for the login page, do you see any usage for that?
in addition, some of those JS yields can contain code related to webpack when loading react components from plugin's .erb which means they need all of the above bundled scripts to exist.

File: app/views/layouts/base.html.erb (hunk 3 above)
Comment (Ron-Lavi) [reply app/views/layouts/base.html.erb]: We are not supporting IE at all, but since IE 10 isn't reading this section, we could keep it as is

File: app/views/layouts/base.html.erb (hunk 2 above)
Comment (Ron-Lavi) [reply app/views/layouts/base.html.erb]: true, that would be better, I will use it.

File: app/views/layouts/base.html.erb (hunk 4 above)
Comment (ekohl) [reply app/views/layouts/base.html.erb]: I'd think `webpacked_plugins_with_global_js` and `webpack_dev_server` are useful, since it gives you live reload while you're developing on the login page. As for the other pages: it probably doesn't make a difference, but keeping fewer code paths is better. Every if you have in your code means you technically need to do more testing so I'm always hesitant with it. Especially since you're breaking basic assumptions pages could rely on previously, potentially making a breaking change for plugins.

So at the very least, move the duplicate code that's in both branches outside of the condition but really I'd keep this as compatible as possible and move the existing code verbatim to the body (just as the PR description says).

**Review by ekohl — DISMISSED** [review]
I think this should be good to merge now. Anyone else wants to weigh in or shall I hit the big green button?

**Review by ShimShtein — DISMISSED** [review]

**Comment by Ron-Lavi** [comment]
Rebased and resolved conflict, @MariaAga any other thoughts on this implementation?

**Review by MariaAga — APPROVED** [review]
Code looks fine to me

**Comment by ekohl** [comment]
Let's get this in, even though we're branching today. It should be an easy revert if it leads to problems.
